File: trip-service/src/main/java/com/study/trip/application/service/TripService.java
package com.study.trip.application.service;

import com.study.trip.application.dto.CreateTripRequest;
import com.study.trip.application.dto.TripResponse;
import com.study.trip.domain.entity.Trip;
import com.study.trip.domain.repository.TripRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.kafka.core.KafkaTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class TripService {

    private final TripRepository tripRepository;
    private final KafkaTemplate<String, String> kafkaTemplate;

    @Transactional
    public TripResponse createTrip(CreateTripRequest request) {
        Trip trip = Trip.builder()
                .userId(request.getUserId())
                .title(request.getTitle())
                .description(request.getDescription())
                .startDate(request.getStartDate())
                .endDate(request.getEndDate())
                .build();
        Trip saved = tripRepository.save(trip);
        kafkaTemplate.send("trip-events", "TRIP_CREATED:" + saved.getId());
        return TripResponse.from(saved);
    }

    public TripResponse getTrip(Long id) {
        Trip trip = tripRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Trip not found: " + id));
        return TripResponse.from(trip);
    }

    public List<TripResponse> getTripsByUser(Long userId) {
        return tripRepository.findByUserId(userId).stream()
                .map(TripResponse::from)
                .toList();
    }
}

File: trip-service/src/main/kotlin/com/study/trip/application/service/TripService.kt
package com.study.trip.application.service

import com.study.common.event.TripCreatedEvent
import com.study.common.exception.EntityNotFoundException
import com.study.common.outbox.OutboxEventPublisher
import com.study.trip.application.dto.CreateTripRequest
import com.study.trip.application.dto.TripResponse
import com.study.trip.domain.entity.Trip
import com.study.trip.domain.repository.TripQueryRepository
import com.study.trip.domain.repository.TripRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class TripService(
    private val tripRepository: TripRepository,
    private val tripQueryRepository: TripQueryRepository,
    private val outboxEventPublisher: OutboxEventPublisher
) {
    @Transactional
    fun createTrip(userId: Long, request: CreateTripRequest): TripResponse {
        val startDate = requireNotNull(request.startDate) { "Start date is required" }
        val endDate = requireNotNull(request.endDate) { "End date is required" }
        val trip = Trip(
            userId = userId, title = request.title, description = request.description,
            startDate = startDate, endDate = endDate
        )
        val saved = tripRepository.save(trip)
        outboxEventPublisher.publish("trip-events", TripCreatedEvent(saved.id, saved.userId, saved.title))
        return TripResponse.from(saved)
    }

    fun getTrip(id: Long): TripResponse {
        val trip = tripRepository.findById(id).orElseThrow { EntityNotFoundException("Trip", id) }
        return TripResponse.from(trip)
    }

    fun getTripsByUser(userId: Long): List<TripResponse> =
        tripRepository.findByUserId(userId).map { TripResponse.from(it) }

    fun searchTrips(
        userId: Long?, status: com.study.trip.domain.entity.TripStatus?,
        keyword: String?, startDateFrom: java.time.LocalDate?, startDateTo: java.time.LocalDate?
    ): List<TripResponse> =
        tripQueryRepository.searchTrips(userId, status, keyword, startDateFrom, startDateTo)
            .map { TripResponse.from(it) }
}
